### ISSUE TITLE
Fine tuning on specific arabic dataset

### ISSUE THREAD

**Issue by omarkshoaib** [edwko/OuteTTS#76]
Hello 
I have been trying your model for a while now and i was attempting to finetune it but i think i am doing something wrong can anyone who got results help steps

 I wanted  to finetune this model on my dataset which is an arabic dataset?

**Comment by edwko** [comment]
At the moment, there’s no official fine-tuning guide for v1, the existing guide is for v0.3. The general process is similar, but fine-tuning v1 does require different data processing scripts to prepare everything properly.

**Comment by omarkshoaib** [comment]
what are the different data processing steps please


i have implemented this but the output is either empty or hallucinates with English words or wrong arabic pronunciation 
am i doing something wrong in the parquet file
the data is clean with no background noises.

`
import os
import csv
import polars as pl
import torchaudio
import torch
from tqdm.notebook import tqdm # Use notebook version for better display
from loguru import logger

from outetts.wav_tokenizer.audio_codec import AudioCodec
from outetts.version.v2.alignment import CTCForcedAlignment
from outetts.version.v2.prompt_processor import PromptProcessor

class DataPreparation:
    def __init__(self, wavs_dir, metadata_file, save_dir, save_len=5000, model_tokenizer_path="OuteAI/Llama-OuteTTS-1.0-1B"):
        self.wavs_dir = wavs_dir
        self.metadata_file = metadata_file
        self.save_dir = save_dir
        self.save_len = save_len  # Number of samples per Parquet file
        self.data = []
        self.save_id = 0
        self.device = "cuda" if torch.cuda.is_available() else "cpu"
        logger.info(f"DataPreparation using device: {self.device}")
        self.audio_codec = AudioCodec(device=self.device, load_decoder=False)
        self.prompt_processor = PromptProcessor(model_tokenizer_path)
        self.ctc = CTCForcedAlignment(self.device)

    def create_speaker(self, wav_path, transcript):
        waveform, sr = torchaudio.load(wav_path)
        if waveform.shape[0] > 1:
            waveform = waveform.mean(dim=0, keepdim=True)
        transcript_clean = transcript.strip()
        speaker_words = self.ctc.align(wav_path, transcript_clean)
        try:
            concatenated = torch.cat([w["audio"] for w in speaker_words], dim=1)
        except Exception as e:
            logger.error(f"Error concatenating audio for file {wav_path}: {e}")
            # Handle cases with no valid words found by alignment
            if not speaker_words:
                 logger.warning(f"No words aligned for {wav_path}. Skipping.")
                 return None # Indicate failure
            raise e
        converted_audio = self.audio_codec.convert_audio_tensor(
            audio=concatenated,
            sr=self.ctc.sample_rate
        ).to(self.audio_codec.device)
        full_codes = self.audio_codec.encode(converted_audio).tolist()
        data_words = []
        start = 0
        for word in speaker_words:
            end = int(round((word["x1"] / self.ctc.sample_rate) * 75))
            try:
                word_tokens = full_codes[0][0][start:end]
            except Exception:
                word_tokens = [1] # Use a default token if slicing fails
            start = end
            if not word_tokens:
                word_tokens = [1] # Use a default token if empty
            data_words.append({
                "word": word["word"],
                "duration": round(len(word_tokens) / 75, 2),
                "codes": word_tokens
            })
        return {"text": transcript_clean, "words": data_words}

    def process(self):
        if not os.path.exists(self.metadata_file):
            logger.error(f"Metadata file not found: {self.metadata_file}")
            return
        if not os.path.isdir(self.wavs_dir):
            logger.error(f"WAVs directory not found: {self.wavs_dir}")
            return

        with open(self.metadata_file, "r", encoding="utf-8") as f:
            # Handle potential empty lines or lines with incorrect format
            reader = csv.reader(f, delimiter="|")
            rows = [row for row in reader if len(row) >= 2 and row[0].strip() and row[1].strip()]

        total = len(rows)
        logger.info(f"Found {total} valid entries in metadata file.")

        for row in tqdm(rows, total=total, desc="Processing audio files"):
            filename, text = row[0].strip(), row[1].strip()
            # Ensure filename ends with .wav (or adjust if other formats are used)
            if not filename.lower().endswith('.wav'):
                filename += '.wav'
            wav_path = os.path.join(self.wavs_dir, filename)

            if not os.path.exists(wav_path):
                logger.warning(f"WAV file {wav_path} does not exist. Skipping.")
                continue
            try:
                speaker = self.create_speaker(wav_path, text)
                if speaker is None: # Skip if create_speaker failed
                    continue
                prompt = self.prompt_processor.get_training_prompt(speaker)
                self.data.append({"prompt": prompt})
            except Exception as e:
                logger.error(f"Error processing {wav_path}: {e}")
                # Optionally add more robust error handling here

            if len(self.data) >= self.save_len:
                self.save_data()

        if self.data: # Save any remaining data
            self.save_data()

        logger.info("Data preparation finished. Freeing resources.")
        self.ctc.free()

    def save_data(self):
        os.makedirs(self.save_dir, exist_ok=True)
        save_path = os.path.join(self.save_dir, f"{self.save_id:06d}.parquet")
        logger.info(f"Saving batch {self.save_id} ({len(self.data)} items) to {save_path}")
        try:
            pl.DataFrame(self.data).write_parquet(save_path)
            self.data = []
            self.save_id += 1
        except Exception as e:
            logger.error(f"Failed to save data to {save_path}: {e}")

# --- Configuration ---
# IMPORTANT: Update these paths to point to your data in Google Drive
DRIVE_BASE_PATH = "/content/drive/MyDrive" # Adjust if your data is elsewhere in Drive
WAVS_DIR = os.path.join(DRIVE_BASE_PATH, "/content/drive/MyDrive/Address/data/wavs") # <-- UPDATE THIS PATH
METADATA_FILE = os.path.join(DRIVE_BASE_PATH, "/content/drive/MyDrive/Address/data/metadata.csv") # <-- UPDATE THIS PATH
SAVE_DIR = "/content/prepared_data" # Save processed data within Colab's temporary storage
SAVE_LEN_PER_FILE = 1000 # Adjust batch size for saving if needed
MODEL_TOKENIZER_PATH = "OuteAI/Llama-OuteTTS-1.0-1B"

# --- Execution ---
logger.info("Starting data preparation...")
dp = DataPreparation(WAVS_DIR, METADATA_FILE, SAVE_DIR, save_len=SAVE_LEN_PER_FILE, model_tokenizer_path=MODEL_TOKENIZER_PATH)
dp.process()
logger.info("Data preparation script finished.")

## 3. Train Model



import glob, os, torch, polars as pl
from loguru import logger
from datasets import Dataset
from transformers import (
    AutoTokenizer, AutoModelForCausalLM,
    Trainer, TrainingArguments,
    DataCollatorForLanguageModeling,
)

# ----------------------------------------------------------------------
# Helper: load prompts stored as *.parquet
# ----------------------------------------------------------------------
def load_dataset_from_parquet(folder: str) -> Dataset:
    prompts = []
    for fpath in glob.glob(os.path.join(folder, "*.parquet")):
        try:
            df = pl.read_parquet(fpath)
            prompts.extend(df["prompt"].to_list())
        except Exception as exc:
            logger.error(f"[load_dataset] {fpath}: {exc}")
    return Dataset.from_dict({"text": prompts})


# ----------------------------------------------------------------------
# Main routine
# ----------------------------------------------------------------------
def main() -> None:
    model_name = "OuteAI/Llama-OuteTTS-1.0-1B"
    data_dir   = "/content/prepared_data"

    tokenizer = AutoTokenizer.from_pretrained(model_name)

    model = AutoModelForCausalLM.from_pretrained(
        model_name,
        device_map="auto",          # fp32 weights (default)
    )
    model.gradient_checkpointing_enable()  # still on


    raw_ds = load_dataset_from_parquet(data_dir)
    raw_ds = raw_ds.train_test_split(test_size=0.1, seed=42)

    def tokenize_fn(batch):
        return tokenizer(
            batch["text"],
            truncation=True,
            max_length=4096,
            padding="max_length",
        )

    tokenised = raw_ds.map(tokenize_fn, batched=True, remove_columns=["text"])
    collator  = DataCollatorForLanguageModeling(tokenizer, mlm=False)

    train_args = TrainingArguments(
        output_dir="/content/fine_tuned_model_2",
        overwrite_output_dir=True,
        num_train_epochs=11,
        per_device_train_batch_size=1,
        per_device_eval_batch_size=1,
        gradient_accumulation_steps=8,   # effective batch 1 × 8 = 8
        gradient_checkpointing=True,     # redundant with .enable(), safe
        eval_strategy="epoch",
        save_strategy="epoch",
        logging_steps=10,
        learning_rate=5e-5,
        weight_decay=0.01,
        remove_unused_columns=False,
        fp16=torch.cuda.is_available(),
        report_to="none",
        save_total_limit=1,
    )

    trainer = Trainer(
        model=model,
        args=train_args,
        train_dataset=tokenised["train"],
        eval_dataset=tokenised["test"],
        data_collator=collator,
    )

    trainer.train()

    model.save_pretrained("/content/fine_tuned_model_2")
    tokenizer.save_pretrained("/content/fine_tuned_model_2")
    logger.info("✅  Model & tokenizer saved to /content/fine_tuned_model_2")


if __name__ == "__main__":
    main()


## 4. Test Model

import outetts as tts, os, torch

fine_tuned = "/content/fine_tuned_model_2"           # your checkpoint dir
assert os.path.isdir(fine_tuned)

# 1 ─ start with an enum‑based auto config …
cfg = tts.ModelConfig.auto_config(
    model   = tts.Models.VERSION_1_0_SIZE_1B,      # tells it the architecture
    backend = tts.Backend.HF,                      # PyTorch checkpoint loader
)

# 2 ─ …then point BOTH paths to your local weights/tokeniser
cfg.model_path     = fine_tuned
cfg.tokenizer_path = fine_tuned

interface = tts.Interface(config=cfg)

# You can use the built‑in demo voice or load your own JSON
speaker = interface.load_default_speaker("EN-FEMALE-1-NEUTRAL")

gen_cfg = tts.GenerationConfig(
    text="في كمبوند ذا مارك بالتجمع الخامس، تتوفر وحدات سكنية متنوعة للبيع. على سبيل المثال، هناك شقق بمساحات تبدأ من ١٢٠ متر مربع، وتاون هاوس بمساحات تبدأ من ١٨٥ متر مربع، وفيلات مستقلة بمساحات تبدأ من ٢٢٥ متر مربع. الأسعار بتبدأ من ٢٤٬١٠٠٬٠٠٠ جنيه مصري. لو عايز تعرف تفاصيل أكتر، ممكن تزور موقع عقارماب",
   speaker=speaker,
    generation_type=tts.GenerationType.CHUNKED,
    sampler_config=tts.SamplerConfig(
        temperature=0.7,
        top_p=0.9,
        repetition_penalty=1.03,   # ← valid field
    ),
)

wav = interface.generate(config=gen_cfg)
wav.save("/content/test_output_interface_15.wav")
print("✓ saved /content/test_output_interface_11.wav")
`

**Comment by edwko** [comment]
This won't work. As mentioned, this script is only compatible with v0.3 models. You're currently trying to use it with a v1.0 model, which has a completely different approach.

For example, the v1.0 model uses `DAC` instead of `WasTokenizer`, and it follows the [v3 interface](https://github.com/edwko/OuteTTS/tree/main/outetts/version/v3) format. Also `CTCForcedAlignment` is not supported in the new method.

I plan to add an example for fine-tuning v1.0 in future updates.

**Comment by omarkshoaib** [comment]
I appreciate your help and efforts truly incredible,

may i ask if you can guide me through what changes should i make to make it work? i would like to contribute to making the fine tuning examples.

**Comment by edwko** [comment]
You’ll need to adjust the script to work with the [interface version v3](https://github.com/edwko/OuteTTS/tree/main/outetts/version/v3), as function names are mostly cross-compatible, so this part should be fairly straightforward. For data creation, you'll need to modify the speaker creation code. In this case, using Whisper to generate the dataset is likely the most straightforward approach. You'll want to retrieve the speaker data through the speaker creation process, but make sure to adjust it so that the Whisper model stays loaded, currently, it gets unloaded after the profile is created.

Refer to these parts for the speaker code:
https://github.com/edwko/OuteTTS/blob/94e25d4f7fde64d416248c3394c9604321b75855/outetts/version/interface.py#L72-L79
https://github.com/edwko/OuteTTS/blob/94e25d4f7fde64d416248c3394c9604321b75855/outetts/whisper/transcribe.py#L12-L18

**Comment by cod3r0k** [comment]
Hi, could you guide me on how to fine-tune a model using a dataset? @omarkshoaib @edwko @haixuanTao @xenova @Ednaordinary

**Comment by edwko** [comment]
@cod3r0k You can use Unsloth guide: https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Oute_TTS_(1B).ipynb
